samples/bluetooth/hci_rpmsg/src/main.c

```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
 	} while (ret < 0);
 
 	LOG_INF("Sent message of %d bytes.", ret);
+	printk("Sent message of %d bytes.", ret);
 
 	net_buf_unref(buf);
 }
 
+#if defined(CONFIG_BT_CTLR_ASSERT_HANDLER) || defined(CONFIG_VS_HCI_SYSTEM_FATAL_ERROR_HANDLER)
+/* A memory pool debug purposes. Currently used sent HCI vendor specific event with LL_ASSERT
+ * information.
+ */
+NET_BUF_POOL_FIXED_DEFINE(debug_tx_pool, 1, BT_BUF_RX_SIZE, 8, NULL);
+
+static struct net_buf *hci_debug_evt_create(uint8_t evt, uint8_t len)
+{
+	struct bt_hci_evt_hdr *hdr;
+	struct net_buf *buf;
+
+	buf = net_buf_alloc(&debug_tx_pool, K_FOREVER);
+	if (buf) {
+		net_buf_reserve(buf, BT_BUF_RESERVE);
+		bt_buf_set_type(buf, BT_BUF_EVT);
+
+		hdr = net_buf_add(buf, sizeof(*hdr));
+		hdr->evt = evt;
+		hdr->len = len;
+	}
+
+	return buf;
+}
+
+#endif /* CONFIG_BT_CTLR_ASSERT_HANDLER || CONFIG_VS_HCI_SYSTEM_FATAL_ERROR_HANDLER */
+
 #if defined(CONFIG_BT_CTLR_ASSERT_HANDLER)
 void bt_ctlr_assert_handle(char *file, uint32_t line)
 {
-	BT_ASSERT_MSG(false, "Controller assert in: %s at %d", file, line);
+	uint32_t len = 0U, pos = 0U;
+
+	/* Disable interrupts, this is unrecoverable */
+	(void)irq_lock();
+
+	if (file) {
+		while (file[len] != '\0') {
+			if (file[len] == '/') {
+				pos = len + 1;
+			}
+			len++;
+		}
+		file += pos;
+		len -= pos;
+	}
+
+	/* Total length: 0xAA + strlen + \0 + 32-bit line number */
+	len = 1 + len + 1 + 4;
+
+	/* Prepare vendor specific HCI debug event */
+	struct net_buf *buf;
+
+	buf = hci_debug_evt_create(BT_HCI_EVT_VENDOR, len);
+	if (!buf) {
+		BT_ASSERT_MSG(false, "Controller assert in: %s at %d", file, line);
+	}
+
+	net_buf_add_u8(buf, 0xAA);
+
+	if (len) {
+		while (*file != '\0') {
+			net_buf_add_u8(buf, *file);
+			file++;
+		}
+		net_buf_add_u8(buf, 0x00);
+	}
+
+	net_buf_add_u8(buf, line >> 0 & 0xff);
+	net_buf_add_u8(buf, line >> 8 & 0xff);
+	net_buf_add_u8(buf, line >> 16 & 0xff);
+	net_buf_add_u8(buf, line >> 24 & 0xff);
+
+	/* Send the event over rpmsg */
+	hci_rpmsg_send(buf);
+
+	while (1) {
+	}
 }
 #endif /* CONFIG_BT_CTLR_ASSERT_HANDLER */
 
+#if defined(CONFIG_VS_HCI_SYSTEM_FATAL_ERROR_HANDLER)
+void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *esf)
+{
+	/* 0xAA + r0/a1 + 0xAA + r1/a2 + 0xAA + r2/a3 + 0xAA + r3/a4 + 0xAA + r12/ip +
+	 * 0xAA + r14/lp + 0xAA + XPSR
+	 */
```

Comment:
Why do you need `0xAA` before every register value?  Use `0x55` or something else a new subevent code for the vendor specific HCI event.